ports/espressif/common-hal/espulp/ULP.c

```diff
@@ -67,13 +76,25 @@ void common_hal_espulp_ulp_run(espulp_ulp_obj_t *self, uint32_t *program, size_t
     }
     pins_used = pin_mask;
 
-    ulp_set_wakeup_period(0, 20000);
+    // Main purpose of ULP is to run while main cpu is in deep sleep, so
+    // ensure GPIO Power Domain remains enabled during deep sleep,
+    // if any GPIO were supplied here.
+    if (pins_used > 0) {
+        esp_sleep_pd_config(ESP_PD_DOMAIN_RTC_PERIPH, ESP_PD_OPTION_ON);
+    }
 
+    int _errno;
     switch (self->arch) {
         #ifdef CONFIG_ULP_COPROC_TYPE_FSM
         case FSM:
-            ulp_load_binary(0, (const uint8_t *)program, length);
-            ulp_run(0);
+            _errno = ulp_load_binary(0, (const uint8_t *)program, length / sizeof(uint32_t));
+            if (_errno != ESP_OK) {
+                mp_raise_RuntimeError(MP_ERROR_TEXT("Load binary failed"));
+            }
+            _errno = ulp_run(entry_point / sizeof(uint32_t));
+            if (_errno != ESP_OK) {
+                mp_raise_RuntimeError(MP_ERROR_TEXT("Run binary failed"));
+            }
```

Comment:
These errors aren't very helpful because they don't suggest what to fix. Either add more specific checks before this to give an error message (and try to reuse an existing message) or use `CHECK_ESP_RESULT()` to check the output:
https://github.com/adafruit/circuitpython/blob/c47f013b9f7596fdaec67a71a891c96ec1f9df84/ports/espressif/bindings/espidf/__init__.h#L21

This maps ESP-IDF errors to Python errors: 
https://github.com/adafruit/circuitpython/blob/c47f013b9f7596fdaec67a71a891c96ec1f9df84/ports/espressif/common-hal/espidf/__init__.c#L57-L121

Reply:
And here I also chose `CHECK_ESP_RESULT` for the `ulp_*_run()` functions, since (to my understanding of the possible failure modes of these functions) there is nothing to user can do in the case that these functions fail. Hence it doesn't make sense to waste more memory for error messages explaining what exactly went wrong.

For the `ulp_*_load_binary()` functions on the other hand, `CHECK_ESP_RESULT` would raise a rather confusing `ESP_ERR_NOT_SUPPORTED` in case of an invalid binary. So instead I reused the existing `mp_raise_ValueError_varg(MP_ERROR_TEXT("Invalid %q"), MP_QSTR_program)`, i.e. `Invalid program`, which might suggest the user to fix the program binary.